### ISSUE TITLE
Cannot sign requests with role based credentials

### ISSUE THREAD

**Issue by eirc** [aws/aws-sdk-ruby#2550]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though [Developer Guide for v3](https://docs.aws.amazon.com/sdk-for-ruby/v3/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/index.html)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-ruby) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-ruby/issues) and didn't find any solution
  
**Describe the bug**
The V4 Signer cannot sign requests when credentials are role-based.

**Gem name ('aws-sdk', 'aws-sdk-resources' or service gems like 'aws-sdk-s3') and its version**
aws-sigv4

**Version of Ruby, OS environment**
ruby 3.0.1p64 (2021-04-05 revision 0fb782ee38) [x86_64-linux]

**To Reproduce (observed behavior)**
`~/.aws/credentials`:
```
[raw]
aws_access_key_id = xxx
aws_secret_access_key = xxx

[role]
role_arn = abc
```

`test.rb`:
```
#!/usr/bin/env ruby

require 'aws-sdk-core'
require 'aws-sigv4'

raw_signer = Aws::Sigv4::Signer.new service: 's3',
  region: 'eu-central-1',
  credentials_provider: Aws::SharedCredentials.new(profile_name: 'raw')

p raw_signer.sign_request url: 'https://domain.com', http_method: 'POST'

raw_signer = Aws::Sigv4::Signer.new service: 's3',
  region: 'eu-central-1',
  credentials_provider: Aws::SharedCredentials.new(profile_name: 'role')

p raw_signer.sign_request url: 'https://domain.com', http_method: 'POST'
```

The first signature is completed successfully but the second one tries to directly access an access key which is not applicable in this case.

Output:

```
#<Aws::Sigv4::Signature:0x000056516bf40958 @headers={"host"=>"domain.com", "x-amz-date"=>"20210707T155354Z", "x-amz-content-sha256"=>"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "authorization"=>"AWS4-HMAC-SHA256 Credential=xxx/20210707/eu-central-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=71ae55b31375287fbf9c2704de9a2eb3cffb178433c0c22c62afec43e92d80af"}, @string_to_sign="AWS4-HMAC-SHA256\n20210707T155354Z\n20210707/eu-central-1/s3/aws4_request\n9757c874cb9d6bf81d219db909b6414ca9b2c14d7a875400efe66919b1bb576f", @canonical_request="POST\n/\n\nhost:domain.com\nx-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date:20210707T155354Z\n\nhost;x-amz-content-sha256;x-amz-date\ne3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", @content_sha256="e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855">
/usr/local/bundle/gems/aws-sigv4-1.2.3/lib/aws-sigv4/signer.rb:684:in `credentials_set?': undefined method `access_key_id' for nil:NilClass (NoMethodError)
        from /usr/local/bundle/gems/aws-sigv4-1.2.3/lib/aws-sigv4/signer.rb:671:in `fetch_credentials'
        from /usr/local/bundle/gems/aws-sigv4-1.2.3/lib/aws-sigv4/signer.rb:208:in `sign_request'
        from /tmp/test.rb:16:in `<main>'
```

**Expected behavior**
The signer should attempt to sign the request in both cases.

**Comment by eirc** [comment]
Ok digging deeper in this all day and it appears I am mistaken, the SDK is not wrong here.

For reference my issue was that I was trying to use `Aws::SharedCredentials` and such objects to load profiles and roles etc. This is not the correct way. To my understanding these are low level data structures that don't resolve the numerous ways AWS credentials are provided. It seems (to me) this happens with the help of the [CredentialProviderChain](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/credential_provider_chain.rb) which hooks into clients as a seahorse configuration plugin.

So the only way to get credentials to resolve properly is to use any `Client` and just pass the profile there.

**Comment by mullermp** [comment]
Yes, you are correct. It doesn't work the way it seems like it should. We should take this up as a documentation enhancement.

**Comment by eirc** [comment]
@mullermp if so I assume you want the ticket to remain open then?

**Comment by alextwoods** [comment]
@eirc  I pushed doc updates to [SharedCredentials](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SharedCredentials.html#initialize-instance_method).  I also created a draft PR for adding a new ProviderChainCredentials that would use the credentials resolution chain (see #2551) - however, we ultimately decided not to merge it because of the additional complexity and messy dependencies on various plugins from the credentials object that were required to get the credential resolution chain working outside of a service client.  This is definitely an area we will improve in the next major version of the SDK.

I'm going to close this issue since we've done the doc update and done some concrete investigation into what it would take to support.
